src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -0,0 +1,474 @@
+            {
+                _diagnosticIds = diagnosticIds;
+            }
+
+            public async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(CancellationToken cancellationToken)
+            {
+                if (CurrentSolution == null)
+                {
+                    return GetDiagnosticData();
+                }
+
+                DocumentId nullTargetDocumentId = null;
+
+                if (CurrentProjectId != null)
+                {
+                    await AppendDiagnosticsAsync(CurrentProject, nullTargetDocumentId, SpecializedCollections.EmptyEnumerable<DocumentId>(), cancellationToken).ConfigureAwait(false);
+                    return GetDiagnosticData();
+                }
+
+                await AppendDiagnosticsAsync(CurrentSolution, cancellationToken).ConfigureAwait(false);
+                return GetDiagnosticData();
+            }
+
+            protected override bool ShouldIncludeDiagnostic(DiagnosticData diagnostic)
+            {
+                return _diagnosticIds == null || _diagnosticIds.Contains(diagnostic.Id);
+            }
+
+            protected override async Task AppendDiagnosticsAsync(Project project, DocumentId targetDocumentId, IEnumerable<DocumentId> documentIds, CancellationToken cancellationToken)
+            {
+                // when we return cached result, make sure we at least return something that exist in current solution
+                if (project == null)
+                {
+                    return;
+                }
+
+                // get analyzers that are not suppressed.
+                // REVIEW: IsAnalyzerSuppressed call seems can be quite expensive in certain condition. is there any other way to do this?
+                var stateSets = StateManager.GetOrCreateStateSets(project).Where(s => ShouldIncludeStateSet(project, s)).ToImmutableArrayOrEmpty();
+
+                var analyzerDriverOpt = await Owner._compilationManager.CreateAnalyzerDriverAsync(project, stateSets, IncludeSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
+
+                var result = await Owner._executor.GetProjectAnalysisDataAsync(analyzerDriverOpt, project, stateSets, cancellationToken).ConfigureAwait(false);
+
+                foreach (var stateSet in stateSets)
+                {
+                    var analysisResult = result.GetResult(stateSet.Analyzer);
+
+                    foreach (var documentId in documentIds)
+                    {
+                        AppendDiagnostics(GetResult(analysisResult, AnalysisKind.Syntax, documentId));
+                        AppendDiagnostics(GetResult(analysisResult, AnalysisKind.Semantic, documentId));
+                        AppendDiagnostics(GetResult(analysisResult, AnalysisKind.NonLocal, documentId));
+                    }
+
+                    if (targetDocumentId == null)
```

Comment:
can targetDocumentId be non-null and documentIds be non-empty? Its not clear what all valid combinations are possible here.